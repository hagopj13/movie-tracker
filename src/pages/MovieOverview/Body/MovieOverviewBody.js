// @flow
import React from 'react';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import Container from '@material-ui/core/Container';

import ActorList from 'components/Actor/List/ActorList';
import ReviewList from 'components/Review/List/ReviewList';
import MovieList from 'components/Movie/List/MovieList';
import type { Movie } from 'types';

import MovieOverviewBodySection from './Section/MovieOverviewBodySection';

type Props = {
  movie: Movie,
};

const useStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(5),
    '& > *': {
      marginBottom: theme.spacing(5),
      '&:last-child': {
        marginBottom: 0,
      },
    },
  },
}));

const MAX_NUMBER_OF_ACTORS = 7;
const MAX_NUMBER_OF_REVIEWS = 4;
const MAX_NUMBER_OF_RECOMMENDATIONS = 10;

const MovieOverviewBody = (props: Props) => {
  const { movie } = props;

  const classes = useStyles();
  const theme = useTheme();
  const centerRecommendations = useMediaQuery(theme.breakpoints.down('xs'));

  const actors = movie.actors?.slice(0, MAX_NUMBER_OF_ACTORS) ?? [];
  const reviews = movie.reviews?.slice(0, MAX_NUMBER_OF_REVIEWS) ?? [];
  const recommendations = movie.recommendations?.slice(0, MAX_NUMBER_OF_RECOMMENDATIONS) ?? [];

  return (
    <Container className={classes.root}>
      <MovieOverviewBodySection title="Top Cast">
        <ActorList actors={actors} />
      </MovieOverviewBodySection>
      <MovieOverviewBodySection title="Top Reviews">
        <ReviewList reviews={reviews} />
      </MovieOverviewBodySection>
      <MovieOverviewBodySection title="Recommendations">
        <MovieList movies={recommendations} center={centerRecommendations} />
      </MovieOverviewBodySection>
    </Container>
  );
};

export default MovieOverviewBody;
