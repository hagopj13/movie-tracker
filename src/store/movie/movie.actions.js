// @flow
import type { Movie, MovieUserState } from 'types';

import MovieActionTypes from './movie.types';

const fetchMovie = (id: string) => ({
  type: MovieActionTypes.FETCH_MOVIE,
  payload: { id },
});

const fetchMovieStart = () => ({
  type: MovieActionTypes.FETCH_MOVIE_START,
});

const fetchMovieSuccess = (movie: Movie, movieUserState: MovieUserState) => ({
  type: MovieActionTypes.FETCH_MOVIE_SUCCESS,
  payload: { movie, movieUserState },
});

const fetchMovieFailure = (error: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_FAILURE,
  payload: { error },
});

const fetchMovieUserState = (id: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE,
  payload: { id },
});

const fetchMovieUserStateStart = () => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_START,
});

const fetchMovieUserStateSuccess = (movieUserState: MovieUserState) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_SUCCESS,
  payload: { movieUserState },
});

const fetchMovieUserStateFailure = (error: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_FAILURE,
  payload: { error },
});

const setIsMovieFavorite = (id: number, isFavorite: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE,
  payload: { id, isFavorite },
});

const setIsMovieFavoriteStart = () => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_START,
});

const setIsMovieFavoriteSuccess = (id: number, isFavorite: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_SUCCESS,
  payload: { id, isFavorite },
});

const setIsMovieFavoriteFailure = (error: string) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_FAILURE,
  payload: { error },
});

const setIsMovieInWatchlist = (id: string, isInWatchlist: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST,
  payload: { id, isInWatchlist },
});

const setIsMovieInWatchlistStart = () => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_START,
});

const setIsMovieInWatchlistSuccess = (id: number, isInWatchlist: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_SUCCESS,
  payload: { id, isInWatchlist },
});

const setIsMovieInWatchlistFailure = (error: string) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_FAILURE,
  payload: { error },
});

const rateMovie = (id: number, rating: number | null) => ({
  type: MovieActionTypes.RATE_MOVIE,
  payload: { id, rating },
});

const rateMovieStart = () => ({
  type: MovieActionTypes.RATE_MOVIE_START,
});

const rateMovieSuccess = (id: number, rating: number | null) => ({
  type: MovieActionTypes.RATE_MOVIE_SUCCESS,
  payload: { id, rating },
});

const rateMovieFailure = (error: string) => ({
  type: MovieActionTypes.RATE_MOVIE_FAILURE,
  payload: { error },
});

export default {
  fetchMovie,
  fetchMovieStart,
  fetchMovieSuccess,
  fetchMovieFailure,
  fetchMovieUserState,
  fetchMovieUserStateStart,
  fetchMovieUserStateSuccess,
  fetchMovieUserStateFailure,
  setIsMovieFavorite,
  setIsMovieFavoriteStart,
  setIsMovieFavoriteSuccess,
  setIsMovieFavoriteFailure,
  setIsMovieInWatchlist,
  setIsMovieInWatchlistStart,
  setIsMovieInWatchlistSuccess,
  setIsMovieInWatchlistFailure,
  rateMovie,
  rateMovieStart,
  rateMovieSuccess,
  rateMovieFailure,
};
