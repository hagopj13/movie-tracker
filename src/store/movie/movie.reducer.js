// @flow
import { handleActions } from 'redux-actions';

import AuthActionTypes from 'store/auth/auth.types';
import type { Movie, MovieUserState } from 'types';

import MovieActionTypes from './movie.types';

type State = {
  details: Movie,
  userState: MovieUserState,
};

export const initialState: State = {
  details: {
    id: null,
    title: '',
    releaseDate: '',
    voteAverage: 0,
    voteCount: 0,
    posterPath: '',
    backdropPath: '',
    tagline: '',
    overview: '',
    runtime: 0,
    status: '',
    genres: [],
    actors: [],
    reviews: [],
    recommendations: [],
  },
  userState: {
    isFavorite: false,
    isInWatchlist: false,
    rating: null,
  },
};

const fetchMovieStart = () => initialState;

const fetchMovieSuccess = (state: State, action) => ({
  ...state,
  details: action.payload.movie,
  userState: action.payload.movieUserState,
});

const fetchMovieUserStateSuccess = (state: State, action) => ({
  ...state,
  userState: action.payload.movieUserState,
});

const setIsMovieFavoriteSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        isFavorite: action.payload.isFavorite,
      },
    };
  }
  return state;
};

const setIsMovieInWatchlistSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        isInWatchlist: action.payload.isInWatchlist,
      },
    };
  }
  return state;
};

const rateMovieSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        rating: action.payload.rating,
      },
    };
  }
  return state;
};

const logoutSuccess = (state: State) => ({
  ...state,
  userState: initialState.userState,
});

const movieActionHandler = {
  [MovieActionTypes.FETCH_MOVIE_START]: fetchMovieStart,
  [MovieActionTypes.FETCH_MOVIE_SUCCESS]: fetchMovieSuccess,
  [MovieActionTypes.FETCH_MOVIE_USER_STATE_SUCCESS]: fetchMovieUserStateSuccess,
  [MovieActionTypes.SET_IS_MOVIE_FAVORITE_SUCCESS]: setIsMovieFavoriteSuccess,
  [MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_SUCCESS]: setIsMovieInWatchlistSuccess,
  [MovieActionTypes.RATE_MOVIE_SUCCESS]: rateMovieSuccess,
  [AuthActionTypes.LOGOUT_SUCCESS]: logoutSuccess,
};

export default handleActions(movieActionHandler, initialState);
