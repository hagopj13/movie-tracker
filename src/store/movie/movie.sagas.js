import { all, call, takeLatest, put, select } from 'redux-saga/effects';

import * as api from 'api/tmdb';
import authSelectors from 'store/auth/auth.selectors';
import { convertResponseToMovie, convertResponseToMovieUserState } from 'api/tmdb/utils';

import MovieActionTypes from './movie.types';
import movieActions from './movie.actions';

export function* fetchMovie({ payload: { id } }) {
  yield put(movieActions.fetchMovieStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    const { data } = yield call(api.getMovieDetails, id, sessionId);
    const movie = yield call(convertResponseToMovie, data);
    const movieUserState = yield call(convertResponseToMovieUserState, data.account_states);
    yield put(movieActions.fetchMovieSuccess(movie, movieUserState));
  } catch (error) {
    yield put(movieActions.fetchMovieFailure(error.status_message));
  }
}

export function* fetchMovieUserState({ payload: { id } }) {
  yield put(movieActions.fetchMovieUserStateStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    const { data } = yield call(api.getMovieUserState, id, sessionId);
    const movieUserState = yield call(convertResponseToMovieUserState, data);
    yield put(movieActions.fetchMovieUserStateSuccess(movieUserState));
  } catch (error) {
    yield put(movieActions.fetchMovieUserStateFailure(error.status_message));
  }
}

export function* setIsMovieFavorite({ payload: { id, isFavorite } }) {
  yield put(movieActions.setIsMovieFavoriteStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  const accountId = yield select(authSelectors.selectAccountId);
  try {
    yield call(api.setIsMovieFavorite, id, isFavorite, sessionId, accountId);
    yield put(movieActions.setIsMovieFavoriteSuccess(id, isFavorite));
  } catch (error) {
    yield put(movieActions.setIsMovieFavoriteFailure(error.status_message));
  }
}

export function* setIsMovieInWatchlist({ payload: { id, isInWatchlist } }) {
  yield put(movieActions.setIsMovieInWatchlistStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  const accountId = yield select(authSelectors.selectAccountId);
  try {
    yield call(api.setIsMovieInWatchlist, id, isInWatchlist, sessionId, accountId);
    yield put(movieActions.setIsMovieInWatchlistSuccess(id, isInWatchlist));
  } catch (error) {
    yield put(movieActions.setIsMovieInWatchlistFailure(error.status_message));
  }
}

export function* rateMovie({ payload: { id, rating } }) {
  yield put(movieActions.rateMovieStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    if (rating) {
      yield call(api.rateMovie, id, rating * 2, sessionId);
    } else {
      yield call(api.deleteMovieRating, id, sessionId);
    }
    yield put(movieActions.rateMovieSuccess(id, rating));
  } catch (error) {
    yield put(movieActions.rateMovieFailure(error.status_message));
  }
}

export function* onFetchMovie() {
  yield takeLatest(MovieActionTypes.FETCH_MOVIE, fetchMovie);
}

export function* onFetchMovieUserState() {
  yield takeLatest(MovieActionTypes.FETCH_MOVIE_USER_STATE, fetchMovieUserState);
}

export function* onSetIsMovieFavorite() {
  yield takeLatest(MovieActionTypes.SET_IS_MOVIE_FAVORITE, setIsMovieFavorite);
}

export function* onSetIsMovieInWatchlist() {
  yield takeLatest(MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST, setIsMovieInWatchlist);
}

export function* onRateMovie() {
  yield takeLatest(MovieActionTypes.RATE_MOVIE, rateMovie);
}

export default function* movieSagas() {
  yield all([
    call(onFetchMovie),
    call(onFetchMovieUserState),
    call(onSetIsMovieFavorite),
    call(onSetIsMovieInWatchlist),
    call(onRateMovie),
  ]);
}
