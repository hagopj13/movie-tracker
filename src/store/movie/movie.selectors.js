// @flow
import { createSelector } from 'reselect';

const selectMovie = (state) => state.movie;

const selectMovieDetails = createSelector([selectMovie], (movie) => movie.details);

const selectMovieUserState = createSelector([selectMovie], (movie) => movie.userState);

export default {
  selectMovieDetails,
  selectMovieUserState,
};
