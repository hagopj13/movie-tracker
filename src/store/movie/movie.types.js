export default {
  FETCH_MOVIE: 'FETCH_MOVIE',
  FETCH_MOVIE_START: 'FETCH_MOVIE_START',
  FETCH_MOVIE_SUCCESS: 'FETCH_MOVIE_SUCCESS',
  FETCH_MOVIE_FAILURE: 'FETCH_MOVIE_FAILURE',
  FETCH_MOVIE_USER_STATE: 'FETCH_MOVIE_USER_STATE',
  FETCH_MOVIE_USER_STATE_START: 'FETCH_MOVIE_USER_STATE_START',
  FETCH_MOVIE_USER_STATE_SUCCESS: 'FETCH_MOVIE_USER_STATE_SUCCESS',
  FETCH_MOVIE_USER_STATE_FAILURE: 'FETCH_MOVIE_USER_STATE_FAILURE',
  SET_IS_MOVIE_FAVORITE: 'SET_IS_MOVIE_FAVORITE',
  SET_IS_MOVIE_FAVORITE_START: 'SET_IS_MOVIE_FAVORITE_START',
  SET_IS_MOVIE_FAVORITE_SUCCESS: 'SET_IS_MOVIE_FAVORITE_SUCCESS',
  SET_IS_MOVIE_FAVORITE_FAILURE: 'SET_IS_MOVIE_FAVORITE_FAILURE',
  SET_IS_MOVIE_IN_WATCHLIST: 'SET_IS_MOVIE_IN_WATCHLIST',
  SET_IS_MOVIE_IN_WATCHLIST_START: 'SET_IS_MOVIE_IN_WATCHLIST_START',
  SET_IS_MOVIE_IN_WATCHLIST_SUCCESS: 'SET_IS_MOVIE_IN_WATCHLIST_SUCCESS',
  SET_IS_MOVIE_IN_WATCHLIST_FAILURE: 'SET_IS_MOVIE_IN_WATCHLIST_FAILURE',
  RATE_MOVIE: 'RATE_MOVIE',
  RATE_MOVIE_START: 'RATE_MOVIE_START',
  RATE_MOVIE_SUCCESS: 'RATE_MOVIE_SUCCESS',
  RATE_MOVIE_FAILURE: 'RATE_MOVIE_FAILURE',
};
