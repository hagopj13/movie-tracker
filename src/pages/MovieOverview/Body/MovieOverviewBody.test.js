import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import { details as movie } from 'store/fixtures/movie';

import MovieOverviewBody from './MovieOverviewBody';

describe('MovieOverviewBody component', () => {
  let shallow;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    const mockProps = {
      movie,
    };
    wrapper = shallow(<MovieOverviewBody {...mockProps} />);
  });

  it('should render the MovieOverviewBody component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
