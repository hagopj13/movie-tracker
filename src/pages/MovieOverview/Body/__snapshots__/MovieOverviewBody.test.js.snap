// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieOverviewBody component should render the MovieOverviewBody component correctly 1`] = `
<WithStyles(ForwardRef(Container))
  className="makeStyles-root-1"
>
  <MovieOverviewBodySection
    title="Top Cast"
  >
    <ActorList
      actors={
        Array [
          Object {
            "character": "Sherlock Holmes",
            "id": 3223,
            "name": "Robert Downey Jr.",
            "profilePath": "/1YjdSym1jTG7xjHSI0yGGWEsw5i.jpg",
          },
          Object {
            "character": "Dr. John Watson",
            "id": 9642,
            "name": "Jude Law",
            "profilePath": "/aQD9e9SPjxCdyqxyXKBmEhNgjtR.jpg",
          },
          Object {
            "character": "Professor James Moriarty",
            "id": 15440,
            "name": "Jared Harris",
            "profilePath": "/mGODbYpkR1NKBOV3rtTfZ7EWoIl.jpg",
          },
        ]
      }
    />
  </MovieOverviewBodySection>
  <MovieOverviewBodySection
    title="Top Reviews"
  >
    <ReviewList
      reviews={
        Array [
          Object {
            "author": "Author Name",
            "content": "As a general adventure and action movie I would consider this as a decent, even good, movie. As a Sherlock Holmes movie…not so much. Unfortunately, watching this movie confirms my suspicions. This is a Hollywood version of Sherlock Holmes. No style, no finesse. A Sherlock Holmes movie should not be a Hollywood action rollercoaster. It should be intelligent, stylish, witty perhaps involving detective work and, above all, British. It didn’t help that they enrolled Stephen Frey. This film is just not right.

I really cannot stand the hostility between Sherlock Holmes and Doctor Watson. There no justification for it in the works of Sir Arthur Conan Doyle. The slow motion scenes are nothing but annoying. The lame jokes are out of place in a Sherlock Holmes movie.

It gets a 5 out of 10 just because it’s actually a decent movie…if you can forget that they claim it to be a Sherlock Holmes one.",
            "id": "5af7018fc3a3681256009e81",
            "url": "https://www.themoviedb.org/review/5af7018fc3a3681256009e81",
          },
        ]
      }
    />
  </MovieOverviewBodySection>
  <MovieOverviewBodySection
    title="Recommendations"
  >
    <MovieList
      center={false}
      movies={
        Array [
          Object {
            "backdropPath": "/kqvEVAvKZpw3MX3hNH6bVK6heBH.jpg",
            "id": 10528,
            "overview": "Eccentric consulting detective, Sherlock Holmes and Doctor John Watson battle to bring down a new nemesis and unravel a deadly plot that could destroy England.",
            "posterPath": "/momkKuWburNTqKBF6ez7rvhYVhE.jpg",
            "releaseDate": "2009-12-23",
            "title": "Sherlock Holmes",
            "voteAverage": 3.6,
            "voteCount": 10235,
          },
          Object {
            "backdropPath": "/bdNHNvZc9fM6Cw3r7HHMCWacI9I.jpg",
            "id": 10764,
            "overview": "Quantum of Solace continues the adventures of James Bond after Casino Royale. Betrayed by Vesper, the woman he loved, 007 fights the urge to make his latest mission personal. Pursuing his determination to uncover the truth, Bond and M interrogate Mr. White, who reveals that the organization that blackmailed Vesper is far more complex and dangerous than anyone had imagined.",
            "posterPath": "/f2rCskvSSr7ywPAnFGhk4FT6Piz.jpg",
            "releaseDate": "2008-10-30",
            "title": "Quantum of Solace",
            "voteAverage": 3.2,
            "voteCount": 5159,
          },
          Object {
            "backdropPath": "/yiggfsFRk5z2005oGojfyKkv7DI.jpg",
            "id": 20526,
            "overview": "Sam Flynn, the tech-savvy and daring son of Kevin Flynn, investigates his father's disappearance and is pulled into The Grid. With the help of a mysterious program named Quorra, Sam quests to stop evil dictator Clu from crossing into the real world.",
            "posterPath": "/xcO8Uun0mfmE3N9CaiCy0tvjpiu.jpg",
            "releaseDate": "2010-12-10",
            "title": "TRON: Legacy",
            "voteAverage": 3.2,
            "voteCount": 4944,
          },
        ]
      }
    />
  </MovieOverviewBodySection>
</WithStyles(ForwardRef(Container))>
`;
